Validation/HGCalValidation/plugins/HGCalValidator.cc

```diff
@@ -192,6 +192,18 @@ void HGCalValidator::dqmAnalyze(const edm::Event& event,
     histoProducerAlgo_->fill_info_histos(histograms.histoProducerAlgo, totallayers_to_monitor_);
   }
 
+  //Consider CaloParticles coming from the hard scatterer, excluding the PU contribution.
+  std::vector<CaloParticle> caloParticlesFromHardScat;
+  for (auto& it_caloPart : caloParticles) {
+    if (it_caloPart.g4Tracks()[0].eventId().event() != 0 or it_caloPart.g4Tracks()[0].eventId().bunchCrossing() != 0) {
+      LogDebug("HGCalValidator") << "Excluding CaloParticles from event: "
+                                 << it_caloPart.g4Tracks()[0].eventId().event()
+                                 << " with BX: " << it_caloPart.g4Tracks()[0].eventId().bunchCrossing() << std::endl;
+      continue;
+    }
+    caloParticlesFromHardScat.push_back(it_caloPart);
```

Comment:
This is making a copy of each hard-scatter caloparticle. Some alternatives:
* store pointers to hard-scatter caloparticles
* store vector<size_t> with a list of hard-scatter indices, and then select only those indices from the original vector
Both of these alternatives have worse memory locality, but reduce memory usage and copying overhead. It would be interesting to see if these improve the performance any further.

Reply:
Hi @kpedro88 , 

Using pointers or the selected indices has far reaching consequences to the whole package since the input to the vast 
majority of the methods in Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc should be changed and subsequently 
a lot of code. Could we please go on with the emplace_back?

Reply:
It's fine for now, but if the memory usage becomes concerning, these alternatives should be revisited.

Reply:
@apsallid move the filter on the CaloParticles where you actually used them, i.e. one level down the stack.
In this way, you do not have to copy anything.

Reply:
@rovere i tested that but for some reason it is really slow. Maybe because we need that check in 5 different places while the code in these places will loop in any case through all CaloParticles which is of the order of 10000. By reducing the collection to ~200 CaloParticles once in the beginning we gain speed. I do not know what else to think.

Reply:
@apsallid doesn't sound like a good explanation. Are you sure, by any chance, you left some part of the code w/o the selection?

Reply:
@apsallid otherwise use the vector of indices down the stack, so that you don't need to loop over and over and make the very same selection.